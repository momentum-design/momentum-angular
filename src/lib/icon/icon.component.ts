import {
  Component,
  EventEmitter,
  HostBinding,
  Input,
  OnInit,
  Output,
  ViewEncapsulation
} from '@angular/core';
import uniqueId from 'lodash-es/uniqueId';
import getColorValue from '@collab-ui/utils/lib/getColorValue';

// const getColorValue = require('@collab-ui/utils/lib/getColorValue');
const iconNames = require('@collab-ui/icons/data/iconNames.json');

@Component({
  selector: 'cui-icon',
  template: `
    <button
      *ngIf="isClickable"
      class="{{ buttonClasses }}"
      [attr.aria-labelledby]="ariaLabelledBy"
      [attr.aria-label]="ariaLabel"
    >
      <i class="{{ classes }}" [style.color]="getColor()"></i>
    </button>
  `,
  styles: [],
  encapsulation: ViewEncapsulation.None,
})
export class IconComponent implements OnInit {
  @Input() public ariaLabel: string;
  @Input() public buttonClassName: string;
  @Input() public color: string;
  @Input() public className: string;
  @Input() public description: string;
  @Input() public name: string;
  @Input() public size: number;
  @Input() public fontSize: number;
  @Input() public title: string;
  @Input() public type: string;

  @Output() click: EventEmitter<any> = new EventEmitter();

  @HostBinding('class') get _class(): string {
    return !this.isClickable ? this.classes : '';
  }
  @HostBinding('style.color') get _color(): string {
    return !this.isClickable ? this.getColor() : '';
  }
  @HostBinding('attr.aria-labelledby') get _ariaLabelledBy(): string {
    return !this.isClickable ? this.ariaLabelledBy : '';
  }
  @HostBinding('attr.aria-label') get _ariaLabel(): string {
    return !this.isClickable ? this.ariaLabel : null;
  }

  public classes: string;
  public buttonClasses: string;
  public titleId: string;
  public descId: string;
  public isClickable: boolean;

  constructor() {}

  ngOnInit() {
    // input validations
    if (!this.name) {
      throw new Error('Attribute "name" is required');
    }
    if (this.type && this.type !== 'white' && this.type !== '') {
      throw new Error('Attribute "type" if defined is either "" or "white"');
    }

    this.isClickable = this.click.observers.length > 0;

    this.classes =
      'cui-icon icon' +
      `${(this.name && ` ${this.getIconNameClass()}`) || ''}` +
      `${(this.className && ` ${this.className}`) || ''}` +
      ``;

    this.buttonClasses =
      'cui-button cui-button--icon' +
      `${(this.type && ` cui-button--icon-${this.type}`) || ''}` +
      `${(this.buttonClassName && ` ${this.buttonClassName}`) || ''}` +
      ``;

    const iconCount = uniqueId();
    this.titleId = `icon-title-${iconCount}`;
    this.descId = this.description ? `icon-desc-${iconCount}` : undefined;
  }

  private getIconNameClass = () => {
    const defaultSize = 16;
    const sizeFromName = Number(this.name.split('_')[1]);

    let nameWithSize = this.name;
    if (sizeFromName && this.fontSize) {
      nameWithSize = `${this.name.split('_')[0]}_${this.fontSize}`;
    }
    if (!sizeFromName) {
      nameWithSize = `${this.name}_${this.size || defaultSize}`;
    }

    const lookupIconName = nameWithSize.replace(/^(icon-)/, '');
    const iconNameClass = `icon-${lookupIconName}`;

    // validate Icon name with Collab Toolkit
    if ((<any>iconNames).indexOf(lookupIconName) < 0) {
      this.consoleHandler('name-error', nameWithSize);
    }

    return iconNameClass;
  };

  public getColor = () => {
    if (!this.color) {
      return 'inherit';
    }

    if (this.color.startsWith('#')) {
      this.consoleHandler('color-warn', this.color);
      return this.color;
    }
    return getColorValue(this.color);
  };

  public get ariaLabelledBy(): string {
    const { ariaLabel, title, description, titleId, descId } = this;

    if (!ariaLabel) {
      if (title && description) {
        return `${titleId} ${descId}`;
      }
      if (title) {
        return `${titleId}`;
      }
      if (description) {
        return `${descId}`;
      }
    }
    return null;
  }

  public consoleHandler = (message, data) => {
    switch (message) {
      case 'color-warn':
        console.warn(
          `[@collab-ui/angular] Icon: ${data} may not exist in the design system,` +
            ` please use a color name from http://collab-ui.cisco.com/styles/colors`
        );
        break;
      case 'name-error':
        console.warn(
          `[@collab-ui/angular] Icon: Icon ${data} does not exist in the design system.` +
            ` Visit https://icons.collab-ui.com for a list of available icons or to request a new icon.`
        );
        break;
    }
  };
}
