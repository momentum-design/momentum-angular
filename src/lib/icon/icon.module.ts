import { NgModule, CUSTOM_ELEMENTS_SCHEMA, ModuleWithProviders } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IconComponent } from './icon.component';

@NgModule({
  imports: [CommonModule],
  declarations: [IconComponent],
  exports: [IconComponent],
  schemas: [CUSTOM_ELEMENTS_SCHEMA]
})
export class IconModule {
  static forRoot(): ModuleWithProviders {
    return {
      ngModule: IconModule,
      providers: [],
    };
  }
}
