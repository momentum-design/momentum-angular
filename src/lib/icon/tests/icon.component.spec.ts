import { async, ComponentFixture, TestBed } from '@angular/core/testing';
import { IconComponent } from '../icon.component';

describe('IconComponent', () => {
  let testComponent: IconComponent;
  let fixture: ComponentFixture<IconComponent>;
  let iconNativeElement: HTMLElement;

  const defaultIconName = 'icon-info_16';

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      declarations: [IconComponent],
    }).compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(IconComponent);
    testComponent = fixture.componentInstance;
  });

  it('should match snapshot', () => {
    testComponent.name = defaultIconName;
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

  describe('general functionality', () => {
    it('should create', () => {
      testComponent.name = defaultIconName;
      fixture.detectChanges();
      expect(testComponent).toBeTruthy();
    });

    it('should have default class name and required name', () => {
      testComponent.name = defaultIconName;
      fixture.detectChanges();

      iconNativeElement = fixture.nativeElement;

      expect(iconNativeElement.className).toEqual('cui-icon icon icon-info_16');
    });
  });

  describe('Test the sizes of <Icon />', () => {
    it(`if Icon name doesn't contain the size, it should include the passed in size prop`, () => {
      const iconName = 'icon-info';
      testComponent.name = iconName;
      testComponent.size = 24;

      fixture.detectChanges();

      iconNativeElement = fixture.nativeElement;

      const expectedIconName = `${iconName}_${testComponent.size}`;
      expect(iconNativeElement.className).toContain(expectedIconName);
    });

    it(`if Icon name doesn't contain size and the size prop was not given, it sets the default size within name`, () => {
      const iconName = 'icon-info';
      const defaultSize = 16;
      testComponent.name = iconName;

      fixture.detectChanges();

      iconNativeElement = fixture.nativeElement;

      const expectedIconName = `${iconName}_${defaultSize}`;
      expect(iconNativeElement.className).toContain(expectedIconName);
    });

    it('should pass in fontSize prop if you want to override an Icon name containing a size', () => {
      const iconName = 'icon-info_16';
      testComponent.name = iconName;
      testComponent.fontSize = 24;

      fixture.detectChanges();

      iconNativeElement = fixture.nativeElement;

      const expectedIconName = 'icon-info_24';
      expect(iconNativeElement.className).toContain(expectedIconName);
    });

    it(`if icon name doesn't exist in toolkit, throw an icon-name error`, () => {
      testComponent.name = 'icon-info_17';
      const consoleHandlerSpy = spyOn(testComponent, 'consoleHandler');
      fixture.detectChanges();

      expect(consoleHandlerSpy).toHaveBeenCalled();
    });
  });

  describe('with click defined', () => {
    beforeEach(async(() => {
      testComponent.click.subscribe(e => {});
      testComponent.name = defaultIconName;
    }));

    it('should render one button element wrapping an i element', () => {
      testComponent.buttonClassName = 'button-class-name';
      fixture.detectChanges();

      iconNativeElement = fixture.nativeElement;
      const button = iconNativeElement.querySelector('button');

      expect(testComponent.isClickable).toBe(true);
      expect(button.className).toContain(
        'cui-button cui-button--icon button-class-name'
      );
      expect(button.childNodes[0].nodeName).toEqual('I');
    });
  });

  describe('Click Type', () => {
    beforeEach(async(() => {
      testComponent.click.subscribe(e => {});
      testComponent.name = defaultIconName;
    }));

    it('default', () => {
      fixture.detectChanges();

      iconNativeElement = fixture.nativeElement;
      const button = iconNativeElement.querySelector('button');

      expect(testComponent.isClickable).toBe(true);
      expect(button.className).not.toContain('cui-button--icon-white');
    });

    it('set to white', () => {
      testComponent.type = 'white';
      fixture.detectChanges();

      iconNativeElement = fixture.nativeElement;
      const button = iconNativeElement.querySelector('button');

      expect(testComponent.isClickable).toBe(true);
      expect(button.className).toContain('cui-button--icon-white');
    });
  });
});
