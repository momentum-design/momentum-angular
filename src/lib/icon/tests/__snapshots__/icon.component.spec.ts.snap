// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IconComponent should match snapshot 1`] = `
<cui-icon
  buttonClasses={[Function String]}
  classes={[Function String]}
  click={[Function EventEmitter]}
  consoleHandler={[Function Function]}
  descId="undefined"
  getAriaLabelledBy={[Function Function]}
  getColor={[Function Function]}
  getColorSpec={[Function Function]}
  getHexFromJSON={[Function Function]}
  getIconNameClass={[Function Function]}
  isClickable="false"
  isolateRoot={[Function Function]}
  name={[Function String]}
  titleId={[Function String]}
>
  
</cui-icon>
`;
